File: android/src/main/kotlin/com/kurban/xue_hua_speaker_earpiece_toggle/AudioRouteController.kt
package com.kurban.xue_hua_speaker_earpiece_toggle

import android.content.Context
import android.media.AudioManager

internal class AudioRouteController(
    context: Context
) {
    private val audioManager =
        context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    private val detector = RouteDetector()
    private val coordinator = SessionCoordinator(audioManager)
    private val applicator = RouteApplicator(audioManager, detector, coordinator)

    var onRouteChanged: ((String) -> Unit)? = null

    fun getRoute(): String = detector.detect(audioManager)

    fun setRoute(route: String): RouteApplyResult {
        val result = applicator.apply(route)
        onRouteChanged?.invoke(detector.detect(audioManager))
        return result
    }

    fun restoreSessionIfNeeded() {
        coordinator.restoreIfNeeded()
    }
}

File: ios/xue_hua_speaker_earpiece_toggle/Sources/xue_hua_speaker_earpiece_toggle/AudioRouteController.swift
import Foundation

final class AudioRouteController {
  private let session: AudioSessionProtocol
  private let detector: RouteDetector
  private let coordinator: SessionCoordinator
  private let applicator: RouteApplicator

  var onRouteChanged: ((String) -> Void)?

  init(session: AudioSessionProtocol = LiveAudioSession()) {
    self.session = session
    let detector = RouteDetector(session: session)
    let coordinator = SessionCoordinator(session: session)
    self.detector = detector
    self.coordinator = coordinator
    self.applicator = RouteApplicator(
      session: session,
      detector: detector,
      coordinator: coordinator
    )
  }

  deinit {
    coordinator.restoreIfNeeded()
  }

  func getRoute() -> String {
    detector.detect()
  }

  func setRoute(_ route: String) throws -> RouteApplyResult {
    let result = try applicator.apply(route: route)
    onRouteChanged?(detector.detect())
    return result
  }

  func restoreSessionIfNeeded() {
    coordinator.restoreIfNeeded()
  }
}
